packages/block-editor/src/components/inserter/utils.js

```diff
@@ -1,45 +1,71 @@
+const getVariationMapper = ( item ) => ( variation ) => ( {
+	...item,
+	id: `${ item.id }-${ variation.name }`,
+	icon: variation.icon || item.icon,
+	title: variation.title || item.title,
+	description: variation.description || item.description,
+	// If `example` is explicitly undefined for the variation, the preview will not be shown.
+	example: variation.hasOwnProperty( 'example' )
+		? variation.example
+		: item.example,
+	initialAttributes: {
+		...item.initialAttributes,
+		...variation.attributes,
+	},
+	innerBlocks: variation.innerBlocks,
+} );
+
 /**
  * Normalizes an inserter block types list and includes variations as separate items.
  *
  * @param {Array} items Denormalized inserter items
+ * @param {number} [limit] Limit returned results by a given threshold. If limit===-1 return all
  * @return {Array} Normalized inserter items.
  */
-export function includeVariationsInInserterItems( items ) {
-	return items.reduce( ( result, item ) => {
-		const { variations = [] } = item;
-		const hasDefaultVariation = variations.some(
-			( { isDefault } ) => isDefault
-		);
+export function includeVariationsInInserterItems( items, limit = -1 ) {
+	const itemsLength = items.length;
+	if ( limit > 0 && itemsLength >= limit ) {
+		// No need to iterate for variations
+		return items.slice( 0, limit );
+	}
+	if ( limit < 1 ) {
+		// Show all available blocks with variations
+		return items.reduce( ( result, item ) => {
+			const { variations = [] } = item;
+			const hasDefaultVariation = variations.some(
+				( { isDefault } ) => isDefault
+			);
 
-		// If there is no default inserter variation provided,
-		// then default block type is displayed.
-		if ( ! hasDefaultVariation ) {
-			result.push( item );
-		}
+			// If there is no default inserter variation provided,
+			// then default block type is displayed.
+			if ( ! hasDefaultVariation ) {
+				result.push( item );
+			}
 
-		if ( variations.length ) {
-			result = result.concat(
-				variations.map( ( variation ) => {
-					return {
-						...item,
-						id: `${ item.id }-${ variation.name }`,
-						icon: variation.icon || item.icon,
-						title: variation.title || item.title,
-						description: variation.description || item.description,
-						// If `example` is explicitly undefined for the variation, the preview will not be shown.
-						example: variation.hasOwnProperty( 'example' )
-							? variation.example
-							: item.example,
-						initialAttributes: {
-							...item.initialAttributes,
-							...variation.attributes,
-						},
-						innerBlocks: variation.innerBlocks,
-					};
-				} )
+			if ( variations.length ) {
+				const variationMapper = getVariationMapper( item );
+				result = result.concat( variations.map( variationMapper ) );
+			}
+
+			return result;
+		}, [] );
+	}
+
+	// Limit is bigger than items length so we can add variations if exist
+	let itemsToAdd = limit - itemsLength;
+	let variationsToAdd = [];
+	for ( const item of items ) {
```

Comment:
In practice, how likely are we to hit this scenario?

The solution chosen here is to basically backfill `items` in no "meaningful" order (just iterating on `items`) until there happen to be enough variations added to `items` that we can break out and serve the result. Which isn't really helpful to users, IMO (but, also, I'm not sure when we will ever have a `limit` so high that we ever need to backfill).

So I recommend the following:

* For the purposes of this PR and fixing the Quick Inserter issue, drop this part entirely (just keep the essential logic around `items.slice( 0, limit )`).

* For a "proper" long-term solution, we should fix this (at the selector level?) so that variations usage is counted and variations can be properly ordered in usage along with the rest of the inserter items.

Does that seem reasonable?

Reply:
> In practice, how likely are we to hit this scenario?

I believe `zero` but I had to make a solution to handle it, as I wanted feedback :).

> drop this part entirely (just keep the essential logic around items.slice( 0, limit )).

I agree with you on this. I'll change it.

> For a "proper" long-term solution, we should fix this (at the selector level?) so that variations usage is counted and variations can be properly ordered in usage along with the rest of the inserter items.

Actually the problem lies in the fact that we add every available variation to be used in other parts of inserter, like searching to insert a block. So this happens after the `selector` stage.

A new selector for variations' usage would be one part of the solution and the second step would be to handle the addition of variation better or more gracefully, maybe with some `config` or even extract that logic to be used when needed.